File: dashboard/lib/server/sqlite-query.py
#!/usr/bin/env python3
from __future__ import annotations

import json
import sqlite3
import sys
from pathlib import Path


def fail(msg: str) -> None:
    print(json.dumps({"error": msg}), end="")
    raise SystemExit(1)


if len(sys.argv) != 4:
    fail("usage: sqlite-query.py <db-path> <sql> <params-json>")

DB_PATH = Path(sys.argv[1])
SQL = sys.argv[2]

try:
    params = json.loads(sys.argv[3])
    if not isinstance(params, list):
        fail("params must be JSON array")
except Exception:
    fail("invalid params json")

if not DB_PATH.exists():
    print(json.dumps([]), end="")
    raise SystemExit(0)

try:
    con = sqlite3.connect(str(DB_PATH))
    con.row_factory = sqlite3.Row
    try:
        cur = con.execute(SQL, params)
        rows = [dict(row) for row in cur.fetchall()]
        out = []
        for row in rows:
            fixed: dict[str, object | str | int | float | None] = {}
            for key, value in row.items():
                if isinstance(value, (bytes, bytearray)):
                    fixed[key] = value.decode('utf-8', errors='ignore')
                else:
                    fixed[key] = value
            out.append(fixed)
        print(json.dumps(out))
    finally:
        con.close()
except Exception as exc:
    fail(str(exc))

File: dashboard/lib/server/sqlite-query.mjs
#!/usr/bin/env node
import { existsSync } from 'node:fs';
import { DatabaseSync } from 'node:sqlite';

function fail(message) {
  process.stdout.write(JSON.stringify({ error: message }));
  process.exit(1);
}

if (process.argv.length !== 5) {
  fail('usage: sqlite-query.mjs <db-path> <sql> <params-json>');
}

const [, , dbPath, sql, paramsJson] = process.argv;
let params;
try {
  params = JSON.parse(paramsJson);
  if (!Array.isArray(params)) fail('params must be JSON array');
} catch {
  fail('invalid params json');
}

if (!existsSync(dbPath)) {
  process.stdout.write(JSON.stringify([]));
  process.exit(0);
}

function normalizeValue(value) {
  if (value instanceof Uint8Array) {
    return Buffer.from(value).toString('utf8');
  }
  return value;
}

try {
  const db = new DatabaseSync(dbPath, { readOnly: true });
  try {
    const stmt = db.prepare(sql);
    const rows = stmt.all(...params).map((row) => {
      const fixed = {};
      for (const [key, value] of Object.entries(row)) {
        fixed[key] = normalizeValue(value);
      }
      return fixed;
    });
    process.stdout.write(JSON.stringify(rows));
  } finally {
    db.close();
  }
} catch (error) {
  fail(error instanceof Error ? error.message : String(error));
}
